src/client/dfs/dfs.c

```diff
@@ -3900,6 +4007,128 @@ dfs_chmod(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode)
 	return rc;
 }
 
+int
+dfs_chown(dfs_t *dfs, dfs_obj_t *parent, const char *name, uid_t uid, gid_t gid, int flags)
+{
+	daos_handle_t		oh;
+	daos_handle_t		th = DAOS_TX_NONE;
+	bool			exists;
+	struct dfs_entry	entry = {0};
+	daos_key_t		dkey;
+	d_sg_list_t		sgl;
+	d_iov_t			sg_iovs[2];
+	daos_iod_t		iod;
+	daos_recx_t		recx[2];
+	size_t			len;
+	dfs_obj_t		*sym;
+	const char		*entry_name;
+	int			i;
+	int			rc;
+
+	if (dfs == NULL || !dfs->mounted)
+		return EINVAL;
+	if (dfs->layout_v <= 2) {
+		D_ERROR("DFS Container has a layout version that does not support dfs_chown()\n");
+		return ENOTSUP;
+	}
+	if (dfs->amode != O_RDWR)
+		return EPERM;
+	if (parent == NULL)
+		parent = &dfs->root;
+	else if (!S_ISDIR(parent->mode))
+		return ENOTDIR;
+	if (name == NULL) {
+		if (strcmp(parent->name, "/") != 0) {
+			D_ERROR("Invalid path %s and entry name is NULL)\n",
+				parent->name);
+			return EINVAL;
+		}
+		name = parent->name;
+		len = strlen(name);
+		oh = dfs->super_oh;
+	} else {
+		rc = check_name(name, &len);
+		if (rc)
+			return rc;
+		oh = parent->oh;
+	}
+
+	/* Check if parent has the entry */
+	rc = fetch_entry(dfs->layout_v, oh, DAOS_TX_NONE, name, len, true, &exists, &entry,
+			 0, NULL, NULL, NULL);
+	if (rc)
+		D_GOTO(out, rc);
+
+	if (!exists)
+		D_GOTO(out, rc = ENOENT);
+
+	if (uid == -1 && gid == -1)
+		D_GOTO(out, rc = 0);
+
+	/** resolve symlink */
+	if (!(flags & O_NOFOLLOW) && S_ISLNK(entry.mode)) {
+		D_ASSERT(entry.value);
+		rc = lookup_rel_path(dfs, parent, entry.value, O_RDWR, &sym, NULL, NULL, 0);
+		if (rc) {
+			D_ERROR("Failed to lookup symlink %s\n", entry.value);
```

Comment:
```suggestion
			D_ERROR("Failed to lookup symlink %s:" DF_RC"\n", entry.value, DP_RC(rc));
```

Reply:
does that work with errno?

Reply:
Ah yes, good point.  No, it should use strerror then.  If the fs copy/symlink patch is able to land before this one then that'll add fault injection testing which should cover this anyway.

Reply:
I still think you need to update the output here, and that once we land the fs copy PR this will start creating test failures if you don't.
```suggestion
			D_ERROR("Failed to lookup symlink '%s': %d (%s)\n", entry.value, rc, strerror(rc));
```